File: Dynamic-Programming/309-M-Best-Time-to-Buy-and-Sell-Stock-with-Cooldown/maximizing_stock_profit_with_cooldown.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        n = len(prices)
        if n <= 1 :
            return 0
        
        hold = -prices[0]
        sold = 0
        rest = 0

        for i in range(1, n):
            next_hold = max(hold, rest - prices[i])
            next_sold = hold + prices[i]
            next_rest = max(rest, sold)

            hold = next_hold
            sold = next_sold
            rest = next_rest
        return max(sold, rest)

File: Dynamic-Programming/309-M-Best-Time-to-Buy-and-Sell-Stock-with-Cooldown/maximizing_stock_profit_with_cooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int len = prices.size();
        if ( len <= 1)
            return 0;

        int hold = -prices[0];
        int sold = 0;
        int rest = 0;

        for (int i=1; i<len; i++) {
            int next_hold = max(hold, rest - prices[i]);
            int next_sold = hold + prices[i];
            int next_rest = max(rest, sold);

            hold = next_hold;
            sold = next_sold;
            rest = next_rest;
        }

        return max(sold, rest);
    }
};